### ISSUE TITLE
Data source cannot compute values for a new list item

### ISSUE THREAD

**Issue by jwvanhollebeke** [hashicorp/terraform#27013]
### Terraform Version

```
Terraform v0.13.5
```

### Terraform Configuration Files

```terraform
locals {
  service_names = [
    "foo",
    // Uncomment the line below after initial "terraform apply"
    //"bar"
  ]

  services = {
    for service, service_name in local.service_names :
    service_name => {
      option01 = "${service_name}-config-option-01"
    }
  }
}

resource "null_resource" "dependency" {
  for_each = local.services

  triggers = {
    setting = each.value.option01
  }
}

data "null_data_source" "data" {
  for_each = local.services

  inputs = {
    id       = null_resource.dependency[each.key].id
    computed = "prefix-${each.value.option01}-suffix"
  }
}

resource "null_resource" "user" {
  for_each = local.services

  triggers = {
    option = each.value.option01
    data   = data.null_data_source.data[each.key].outputs["computed"]
  }
}
```

### Debug Output

https://gist.github.com/jwvanhollebeke/4f9dc3e5f88ae3c2385509f5c74489fb

### Crash Output

N/A

### Expected Behavior

We manage infrastructure for services that use our platform with an external data set (represented here by the list and map in the `locals` block). We "enroll" a new service by adding it to the external data set.

When we add a new service entry to the list, we expect that additional resources will be created for that service during `terraform apply`.

### Actual Behavior

When adding a new service, receive the error below during `terraform plan`:

```
Error: Invalid index

  on main.tf line 50, in resource "null_resource" "user":
  50:     data   = data.null_data_source.data[each.key].outputs["computed"]
    |----------------
    | data.null_data_source.data is object with 1 attribute "foo"
    | each.key is "bar"

The given key does not identify an element in this collection value.
```

### Steps to Reproduce

1. `terraform init`
2. `terraform apply`
3. Uncomment line 5 of the Terraform configuration.
3. `terraform plan` (or any command that prompts a state refresh).

### Additional Context

We discovered this issue when we upgraded from Terraform `0.12.26` to `0.13.5`. We verified that we do not receive an error when running this configuration with Terraform `0.12.26`.

**Comment by alisdair** [comment]
Hi @jwvanhollebeke, thanks for reporting this. I was able to reproduce this issue with Terraform 0.13.5.

There was a lot of work done in the 0.14 series to fix bugs with refresh during plan (e.g. #26270). I can confirm that this issue has been fixed as of Terraform 0.14.0-rc1, and the fix will ship as part of Terraform 0.14. As a result I'm going to mark this issue fixed. Unfortunately we won't be able to backport these changes to an 0.13 release, as they're too involved.

Until you're able to upgrade to Terraform 0.14, I think you should be able to work around the issue by using targeted apply on the dependent data sources. For example, changing the reproduction steps as follows works for this configuration:

1. `terraform init`
1. `terraform apply`
1. Uncomment line 5 of the Terraform configuration.
1. 👉 `terraform apply -target=-target=data.null_data_source.data`
1. `terraform plan` (or any command that prompts a state refresh).

Thanks again, and I hope this workaround works for you!

**Comment by jwvanhollebeke** [comment]
Thanks so much @alisdair for confirming about Terraform 0.14! I'll document that we can revert our current workaround once we upgrade to TF 0.14. I appreciate the prompt and thoughtful reply.
